src/Compiler/Driver/CompilerDiagnostics.fs

```diff
@@ -2246,11 +2246,103 @@ type PhasedDiagnostic with
 
     /// used by fsc.exe and fsi.exe, but not by VS
     /// prints error and related errors to the specified StringBuilder
-    member diagnostic.Output(buf, tcConfig: TcConfig, severity) =
+    member diagnostic.Output(buf: StringBuilder, tcConfig: TcConfig, severity: FSharpDiagnosticSeverity) =
 
         // 'true' for "canSuggestNames" is passed last here because we want to report suggestions in fsc.exe and fsi.exe, just not in regular IDE usage.
         let diagnostics = CollectFormattedDiagnostics(tcConfig, severity, diagnostic, true)
 
+        let renderRich (details: FormattedDiagnosticDetailedInfo) =
```

Comment:
If this does not need any environmental information and it only takes the details object, you can also consider extracting this to a separate standalone function and test it in isolation.